src/main/java/org/junitpioneer/jupiter/StdIoExtension.java

```diff
@@ -0,0 +1,117 @@
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+class StdIoExtension implements ParameterResolver, BeforeTestExecutionCallback, AfterEachCallback {
+
+	static final String SEPARATOR = System.getProperty("line.separator");
+
+	private static final Namespace NAMESPACE = Namespace.create(StdIoExtension.class);
+
+	private static final String SYSTEM_IN_KEY = "StdIo_System_In";
+	private static final String STD_IN_KEY = "StdIo_Std_In";
+	private static final String OUT_KEY = "StdIo_Std_Out";
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		Class<?> type = parameterContext.getParameter().getType();
+		return (type == StdIn.class || type == StdOut.class);
+	}
+
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		Class<?> parameterType = parameterContext.getParameter().getType();
+		if (parameterType == StdOut.class)
+			return prepareStdOut(extensionContext);
+		if (parameterType == StdIn.class) {
+			String[] source = extensionContext.getRequiredTestMethod().getAnnotation(StdIo.class).value();
+			if (source.length == 0)
+				throw new ExtensionConfigurationException(
+					"@StdIo defined no input, so System.in is still in place and no StdIn parameter can be provided. "
+							+ "If you want to define empty input, use `@StdIo(\"\")`.");
+			else
+				//@formatter:off
+				return extensionContext
+						.getStore(NAMESPACE)
+						.getOrComputeIfAbsent(
+								STD_IN_KEY,
+								__ -> createSwapStoreStdIn(extensionContext, source),
+								StdIn.class);
+				//@formatter:on
+		}
+		throw new ParameterResolutionException(format("Could not resolve parameter of type %s.", parameterType));
+	}
+
+	private StdOut prepareStdOut(ExtensionContext context) {
+		storeStdOut(context);
+		return createOut();
+	}
+
+	private void storeStdOut(ExtensionContext context) {
+		context.getStore(NAMESPACE).put(OUT_KEY, System.out); //NOSONAR never writing to System.out, only storing it
+	}
+
+	private StdOut createOut() {
+		StdOut out = new StdOut();
+		System.setOut(new PrintStream(out));
+		return out;
+	}
+
+	private StdIn createSwapStoreStdIn(ExtensionContext context, String[] source) {
+		StdIn newStdIn = new StdIn(source);
+		swapAndStoreIn(context, newStdIn);
+		return newStdIn;
+	}
+
+	private void swapAndStoreIn(ExtensionContext context, StdIn stdIn) {
+		context.getStore(NAMESPACE).put(SYSTEM_IN_KEY, System.in); //NOSONAR never reading from System.in, only storing it
+		context.getStore(NAMESPACE).put(STD_IN_KEY, stdIn);
+		System.setIn(stdIn); //NOSONAR required to redirect output
+	}
+
+	@Override
+	public void beforeTestExecution(ExtensionContext context) {
+		final Method method = context.getRequiredTestMethod();
+		if (!method.isAnnotationPresent(StdIo.class))
```

Comment:
The `@StdIo` annotation _must_ be present. The StdIoExtension itself is not public and can only be registered with the annotation.

We could delete this condition and exception because this never happens and could be confusing.

Reply:
I switched annotation detection to our utils (so meta-annotations are found as well), which forces us to consider the absent case because it returns an `Optional`.